core/src/main/java/org/apache/calcite/sql/dialect/PostgresqlSqlDialect.java

```diff
@@ -155,6 +159,17 @@ public PostgresqlSqlDialect(Context context) {
     }
   }
 
+  @Override public boolean supportsImplicitTypeCoercion(RexCall call) {
+    final RexNode operand0 = call.getOperands().get(0);
+    RelDataType callType = call.getType();
+    boolean supportImplicit = super.supportsImplicitTypeCoercion(call);
+    boolean isNumericType = supportImplicit && SqlTypeUtil.isNumeric(callType);
+    if (isNumericType) {
+      return false;
+    }
+    return supportImplicit && RexUtil.isLiteral(operand0, false);
```

Comment:
are you sure only literals are a problem?

Reply:
Yes. Literals don't need the cast, others need it. I add a unit test.